### DISCUSSION TITLE
In Relation Based Edge, how to get Source Node

### DISCUSSION THREAD

**Discussion by pelva** [eclipse-sirius/sirius-desktop#335] [Q&A]
Hi,
My understanding: self is the target semantic element.
As shown in the figure below, there are several available variables, and they all seem to point to a DDiagram.
I want to know what the "diagram" variable below represents. 
Is it the current Relation Based Edge? 
If I want to get the Source Node, what should I do? 
I tried using diagram.sourceNode, but I compared self with diagram.sourceNode.target and found that they are not consistent.
please help me,thank you
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/b136d557-bf89-4f6e-8520-e285e5542df0)

**Comment by lredor** [comment]
You're right, all the variables, `diagram`, `viewpoint` and `viewPoint`, point to the `DDiagram` that you are currently work on. It is not necessarily necessary here.

In the target finder expression, you must start from `aql:self` that is the semantic element of the source of this edge. In other words, you must define the expression which allows, from a source semantic element, to find the semantic elements which will be the targets of this relationship.

If you want, you have an example in [Flow Designer](https://github.com/ObeoNetwork/Flow-Designer/blob/master/plugins/fr.obeo.dsl.designer.sample.flow.design/description/flow.odesign#L345):
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/1825882/e6b07e6a-c789-440d-9d80-f119a83843f7)

**Reply by pelva** [reply]
Flow Designer is a great example and I learned a lot from it.
The Edge I implemented is slightly more complicated, it and the connected endpoint are both calculated, so I want to get the Source Node in order to determine the parent container object. In short, thank you for your reply.

**Reply by pelva** [reply]
I think I know how to implement it. Edge has not been created in General->Target Finder Expression, so I can get the SourceView in Advanced->Precondition Expression for further processing.
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/929d7a1e-c81f-4b1a-8b08-e8d357136d63)
